marlin/src/ahp/matrices.rs

```diff
@@ -103,147 +99,99 @@ pub struct MatrixEvals<F: PrimeField> {
 /// Here `M^*(i, j) := M(j, i) * u_H(j, j)`. For more details, see [\[COS20\]](https://eprint.iacr.org/2019/1076).
-#[derive(Derivative)]
+#[derive(derivative::Derivative)]
 #[derivative(Clone(bound = "F: PrimeField"))]
 #[derive(Debug, CanonicalSerialize, CanonicalDeserialize)]
 pub struct MatrixArithmetization<F: PrimeField> {
     /// LDE of the row indices of M^*.
     pub row: LabeledPolynomial<F>,
     /// LDE of the column indices of M^*.
     pub col: LabeledPolynomial<F>,
-    /// LDE of the non-zero entries of A^*.
-    pub val_a: LabeledPolynomial<F>,
-    /// LDE of the non-zero entries of B^*.
-    pub val_b: LabeledPolynomial<F>,
-    /// LDE of the non-zero entries of C^*.
-    pub val_c: LabeledPolynomial<F>,
-    /// LDE of the vector containing entry-wise products of `row` and `col`,
-    /// where `row` and `col` are as above.
+    /// LDE of the vector containing entry-wise products of `row` and `col`.
     pub row_col: LabeledPolynomial<F>,
+    /// LDE of the non-zero entries of M^*.
+    pub val: LabeledPolynomial<F>,
 
-    /// Evaluation of `self.row`, `self.col`, and `self.val` on the domain `K`.
+    /// Evaluation of `self.row_a`, `self.col_a`, and `self.val_a` on the domain `K`.
     pub evals_on_K: MatrixEvals<F>,
 }
 
 // TODO for debugging: add test that checks result of arithmetize_matrix(M).
 pub(crate) fn arithmetize_matrix<F: PrimeField>(
-    joint_matrix: &[Vec<usize>],
-    a: &Matrix<F>,
-    b: &Matrix<F>,
-    c: &Matrix<F>,
-    interpolation_domain: EvaluationDomain<F>,
-    output_domain: EvaluationDomain<F>,
+    matrix: &Matrix<F>,
+    label: &str,
+    non_zero_domain: EvaluationDomain<F>,
+    constraint_domain: EvaluationDomain<F>,
     input_domain: EvaluationDomain<F>,
 ) -> MatrixArithmetization<F> {
     let matrix_time = start_timer!(|| "Computing row, col, and val LDEs");
 
-    let elems: Vec<_> = output_domain.elements().collect();
+    let elems: Vec<_> = constraint_domain.elements().collect();
 
     let lde_evals_time = start_timer!(|| "Computing row, col and val evals");
+
     // Recall that we are computing the arithmetization of M^*,
     // where `M^*(i, j) := M(j, i) * u_H(j, j)`.
-    let a = a
-        .iter()
-        .enumerate()
-        .map(|(r, row)| row.iter().map(move |(f, i)| ((r, *i), *f)))
-        .flatten()
-        .collect::<BTreeMap<(usize, usize), F>>();
-
-    let b = b
-        .iter()
-        .enumerate()
-        .map(|(r, row)| row.iter().map(move |(f, i)| ((r, *i), *f)))
-        .flatten()
-        .collect::<BTreeMap<(usize, usize), F>>();
-
-    let c = c
-        .iter()
-        .enumerate()
-        .map(|(r, row)| row.iter().map(move |(f, i)| ((r, *i), *f)))
-        .flatten()
-        .collect::<BTreeMap<(usize, usize), F>>();
 
     let eq_poly_vals_time = start_timer!(|| "Precomputing eq_poly_vals");
-    let eq_poly_vals: BTreeMap<F, F> = output_domain
+    let eq_poly_vals: HashMap<F, F> = constraint_domain
```

Comment:
just making sure that the order of elements in this collection can be random - `BTreeMap` enforces key ordering, while `HashMap` doesn't

Reply:
Right, I don't think we're iterating over the contents of the map, only using it as a key-value store, so I think we can use the `HashMap` without issue.